flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java

```diff
@@ -339,12 +338,26 @@ private static Request createRequest(String targetAddress, String targetUrl, Htt
 			.thenComposeAsync(
 				channel -> {
 					ClientHandler handler = channel.pipeline().get(ClientHandler.class);
-					CompletableFuture<JsonResponse> future = handler.getJsonFuture();
+
+					CompletableFuture<JsonResponse> future;
+					boolean success = false;
+
 					try {
-						httpRequest.writeTo(channel);
+						if (handler == null) {
+							throw new IOException("Netty pipeline was not properly initialized.");
+						} else {
+							httpRequest.writeTo(channel);
+							future = handler.getJsonFuture();
+							success = true;
+						}
 					} catch (IOException e) {
-						return FutureUtils.completedExceptionally(new FlinkException("Could not write request.", e));
+						future = FutureUtils.completedExceptionally(new ConnectionException("Could not write request.", e));
+					} finally {
+						if (!success) {
+							channel.close();
```

Comment:
why aren't we closing the channel in both cases? Not being initialized _properly_ doesn't necessarily mean that there's nothing to clean up.

Reply:
We keep the channel open if we managed to send the request to the remote endpoint, because we expect an answer. In all other cases (`!success`) we should close it, because we cannot expect something to return.

Reply:
ah of course 👍